include/swift/AST/TypeCheckRequests.h

```diff
@@ -2020,6 +2025,35 @@ class PatternTypeRequest
   }
 };
 
+/// Type-checks a `@differentiable` attribute and returns the resolved parameter
+/// indices on success. On failure, emits diagnostics and returns `nullptr`.
+///
+/// Currently, this request resolves other `@differentiable` attribute
+/// components but mutates them in place:
+/// - `JVPFunction`
+/// - `VJPFunction`
+/// - `DerivativeGenericSignature`
+class DifferentiableAttributeTypeCheckRequest
+    : public SimpleRequest<DifferentiableAttributeTypeCheckRequest,
+                           IndexSubset *(DifferentiableAttr *),
+                           CacheKind::SeparatelyCached> {
```

Comment:
Using `CacheKind::Cached` seems preferable so that `DifferentiableAttr` is truly stateless, adhering to the [request evaluator caching vision](https://github.com/apple/swift/blob/tensorflow/docs/RequestEvaluator.md#caching).

I tried changing `CacheKind::SeparatelyCached` to `CacheKind::Cached` (some time ago on `tensorflow` branch), but it did not seem to work: `@differentiable` attributes in non-primary-files were left unchecked.

I'm not sure why `CacheKind::Cached` did not work. I did notice that quite a few other requests (e.g. `InterfaceTypeRequest`) used `CacheKind::SeparatelyCached` for some reason - perhaps that's just for request cycle breaking?